### DISCUSSION TITLE
0.5.4: Why isn't PortPrototype an ARElement?

### DISCUSSION THREAD

**Discussion by rcd09** [cogu/autosar#124] [General]
Because its not, I can't add it to a package.  Bug?

**Comment by rcd09** [comment]
Now that I've dug deeper, I see the support for ProvidePortPrototype and RequirePortPrototype in the switcher_non_collectible dict that has a comment "only for unit test purposes".

I use these in my application.  Can I help get them implemented?  What needs to be done to make them writable elements?

**Reply by cogu** [reply]
You should not depend on things in the switcher_non_collectible. That is indeed just for unit testing purposes to allow parsing snippets of XML strings.

**Comment by cogu** [comment]
PortPrototypes always belong to a software component. As far as I know it has never been allowed in the ARXML schema (see for example AUTOSAR_00051.xsd) to add PortProtypes directly to a package. What would a package with ports even mean?

Are you perhaps trying to create some sort of "generic" ports library that you want to "import" later and use in multiple software components? If so, then read on:
With Python autosar v0.4 we always solved this issue using my template mechanism where we created a file called "ports.py" which contained the "recipes" for how to create each port (Its name, port-interface, init-value etc.). Later, when we defined the SWCs we imported "ports.py" and simply applied the port templates. All you had to do was to provide the direction of the port (Send/Provide/Serve or Receive/Require/Call).

It would look something like this:
```python
swc = package.createApplicationSoftwareComponent(componentName)
swc.apply(Modes.EcuM_CurrentMode.Require)
swc.apply(Signals.SWS_PushButtonStatus_Back.Receive)
swc.apply(Signals.SWS_PushButtonStatus_Down.Receive)
swc.apply(Signals.SWS_PushButtonStatus_Enter.Receive)
swc.apply(Signals.SWS_PushButtonStatus_Home.Receive)
swc.apply(Signals.SWS_PushButtonStatus_Left.Receive)
swc.apply(Signals.SWS_PushButtonStatus_Right.Receive)
swc.apply(Signals.SWS_PushButtonStatus_Up.Receive)
swc.apply(Services.BspApi.Call)
```

The code above I have copied from my very old autosar-demo project. See https://github.com/cogu/autosar-demo/blob/master/template/components/SteeringWheelButtonFeedback.py. (In that project I called the ports file Signals.py for some reason).

I haven't yet created an equivalent version of "ports.py" in the template example in v0.5.4. I could try and add it for v0.5.5.

If the above is not what you were looking for and you still want to create packages with ports I can't help you since that would mean dealing with invalid ARXML.

**Comment by rcd09** [comment]
I see.  This is all my incomplete understanding of the structure of AUTOSAR ARXML. Thank you for the explanation.  I'll close this issue and continue working on my port from 0.4 to 0.5 with this info.

Thank you for the explanation.
